blog/_posts/2016-07-04-trait-method-performance.md

```diff
@@ -121,9 +130,19 @@ virtual methods that are not overridden.
 
 When the JVM loads a new class, a virtual method that was statically not overridden by CHA may get
 an override. All assembly code that made use of the now invalid assumption is discarded and the
+// the text reads as if the methods are executed twice. I'd rewrite as `the next execution of method will began in intepreter`
+// this leaves uncovered the question of `what happens if return address is in discarded code` but I believe it's fine.
 corresponding methods are executed again by the bytecode interpreter. This process is called
 deoptimization. In case a deoptimized method is currently being executed, control is passed to
 the interpreter by on-stack replacement.
+// AFAIK this is false... Please tell me if what I write below isn't true:
+// when breaking CHA assumption, VM will ask all threads to pause at safe poins.
+// JVM can recreate state of interpreter in every safe-point. If method happens to pause inside method whose code is
+// being discarded it will jump into interpreter
```

Comment:
"JVM can recreate state of interpreter in every safe-point" - how? by creating the apropriate state / stack frame for the interpreter and passing control to it, right? IUC, this is called OSR, too, not only going to new compiled code.